### PR TITLE
ClusterLoader - Adding resource usage measurement

### PR DESCRIPTION
**Pull request by krzysied**

### CODE REVIEW

File: clusterloader2/pkg/measurement/common/api_responsiveness.go [hunk 1]
```diff
@@ -201,12 +202,12 @@ func getMetrics(c clientset.Interface) (string, error) {
...
-	Subresource string                        `json:"subresource"`
-	Verb        string                        `json:"verb"`
-	Scope       string                        `json:"scope"`
-	Latency     measurementutil.LatencyMetric `json:"latency"`
-	Count       int                           `json:"count"`
+	Resource    string                `json:"resource"`
+	Subresource string                `json:"subresource"`
+	Verb        string                `json:"verb"`
+	Scope       string                `json:"scope"`
+	Latency     metrics.LatencyMetric `json:"latency"`
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/api_responsiveness.go]: Can you please:
1. move the "file movement" to a separate PR (it's polluting this one)
2. in this PR, comment on which files are copied (almost copied) from k8s repo and which are written from scratch?

Comment (krzysied) [reply clusterloader2/pkg/measurement/common/api_responsiveness.go]: 1. Done.
2. Comments added. Almost all of the `/util` code is copied.
I tried to split code into gatherers, models and component accessing categories - that's why there are 5 files even though originally were 2.

File: clusterloader2/pkg/measurement/common/resource_usage.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	}
+	provider, err := util.GetString(config.Params, "provider")
+	if err != nil {
+		return summaries, err
+	}
+	host, err := util.GetStringOrDefault(config.Params, "host", "")
+	if err != nil {
+		return summaries, err
+	}
+	masterOnly, err := util.GetBoolOrDefault(config.Params, "masterOnly", false)
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/resource_usage.go]: In the meantime we changed this one to enum (we added masterAndDNS).
Please reflect this here.

File: clusterloader2/pkg/measurement/util/gatherers/container_resource_gatherer.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+Copied from k8s.io/kubernetes/test/e2e/framework/resource_usage_gatherer.go
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/util/gatherers/container_resource_gatherer.go]: Sorry - I didn't mean to add a comment in the file - I meant github comment to make review easier.
Please remove it, because it will probably soon be outdated.

And everywhere else.

File: clusterloader2/pkg/measurement/util/resource.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
+/*
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/util/resource.go]: Would it be possible to split this file to a separate PR - it seems to be independent and for me review time is exponential from the PR size :)

**Comment by krzysied** [comment]
/hold
Waiting for #240

File: clusterloader2/pkg/measurement/util/resource.go (hunk 4 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/util/resource.go]: File moved to #240

File: clusterloader2/pkg/measurement/util/gatherers/container_resource_gatherer.go (hunk 3 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/util/gatherers/container_resource_gatherer.go]: Comments removed.

File: clusterloader2/pkg/measurement/common/resource_usage.go (hunk 2 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/common/resource_usage.go]: Functions updated to newer version.

**Comment by wojtek-t** [comment]
240 is merged now

**Comment by wojtek-t** [comment]
/hold cancel

please rebase

File: clusterloader2/pkg/measurement/util/kubelet/kubelet.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
+/*
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/util/kubelet/kubelet.go]: Same comment here - this is pretty independent - can we move this one to a separate PR?

Comment (krzysied) [reply clusterloader2/pkg/measurement/util/kubelet/kubelet.go]: Utils for kubelet and kubemark moved to separate PRs - #243 #244 .

**Comment by krzysied** [comment]
/hold cancel
@wojtek-t Should I move the gatherers to separate RP?

**Comment by wojtek-t** [comment]
> @wojtek-t Should I move the gatherers to separate RP?

That's fine - it's not super long now. I will take a look in few minutes.

File: clusterloader2/pkg/measurement/common/resource_usage.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+func createResourceUsageMetricMeasurement() measurement.Measurement {
+	return &resourceUsageMetricMeasurement{}
+}
+
+type resourceUsageMetricMeasurement struct {
+	gatherer *gatherers.ContainerResourceGatherer
+}
+
+// Execute supports two actions:
+// - start - Stsrts resource metrics collecting.
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/resource_usage.go]: Starts

File: clusterloader2/pkg/measurement/util/gatherers/container_resource_gatherer.go [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/pkg/util/system"
+	"k8s.io/perf-tests/clusterloader2/pkg/measurement/util"
+)
+
+// NodesSet is a flag defining the node set range.
+type NodesSet int
+
+const (
+	//AllNodes - all containers on all nodes
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/util/gatherers/container_resource_gatherer.go]: nit: space after //

also below

File: clusterloader2/pkg/measurement/common/resource_usage.go [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+			PrintVerboseLogs:            false,
+		}, nil)
+		if err != nil {
+			return summaries, err
+		}
+		go e.gatherer.StartGatheringData()
+		return summaries, nil
+	case "gather":
+		glog.Infof("%v, gathering resource usage...", resourceUsageMetricName)
+		summary, err := e.gatherer.StopAndSummarize([]int{90, 99, 100}, make(map[string]measurementutil.ResourceConstraint))
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/resource_usage.go]: For now there isn't any ability to pass those constraint.
I'm fine with doing it in a separate PR, but this needs to be solved to migrate existing tests.

File: clusterloader2/pkg/measurement/util/gatherers/resource_gather_worker.go [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+			case <-time.After(w.resourceDataGatheringPeriod):
+				w.singleProbe()
+			case <-w.stopCh:
+				return
+			}
+		}
+	case <-w.stopCh:
+		return
+	}
+}
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/util/gatherers/resource_gather_worker.go]: The two files (gatherer) modulo minor nits seem fine, so if you prefer splitting them into separate PR and have them merged faster, I'm fine with that.

File: clusterloader2/pkg/measurement/common/resource_usage.go (hunk 6 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/common/resource_usage.go]: Fixed.

File: clusterloader2/pkg/measurement/common/resource_usage.go (hunk 8 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/common/resource_usage.go]: TODO added - I'll do this separate PR.
Issue added - #248
